linalg/solvers.cpp

```diff
@@ -734,7 +735,7 @@ void CGSolver::Mult(const Vector &b, Vector &x) const
    {
       d = r;
    }
-   nom0 = nom = Dot(d, r);
+   nom0 = nom = initial_norm = Dot(d, r);
```

Comment:
We need to take a square root here; also handle the case when the preconditioner is not PD and the dot product can be negative:
```suggestion
   nom0 = nom = Dot(d, r);
   initial_norm = nom0 >= 0.0 ? sqrt(nom0) : -sqrt(-nom0);
```

Reply:
See ce741763556d316da12702e62a945561787d680e